### PR TITLE
grandpa: only send catch-up requests to authorities

### PR DESCRIPTION
**Pull request by andresilva**
Observers update their view with neighbor packets but since they don't run the full GRANDPA protocol they don't keep track of the vote data necessary to reply to catch-up requests. We need to make sure that we only issue catch up requests to authorities.

### CODE REVIEW

File: core/finality-grandpa/src/communication/gossip.rs [hunk 1]
```diff
@@ -1794,4 +1799,55 @@ mod tests {
...
+				},
+			);
+
+			catch_up_request
+		};
+
+		// importing a neighbor message from a peer in the same set in a later
+		// round should lead to a catch up request but since the node is not an
+		// authority we should get `None`.
+		match import_neighbor_message(peer_full) {
```
Comment (bkchr) [inline core/finality-grandpa/src/communication/gossip.rs]:
```suggestion
		if import_neighbor_message(peer_full).is_some() {
```

File: core/finality-grandpa/src/communication/gossip.rs [hunk 2]
```diff
@@ -1794,4 +1799,55 @@ mod tests {
...
+			None => {},
```
Comment (bkchr) [inline core/finality-grandpa/src/communication/gossip.rs]:
```suggestion
```

File: core/finality-grandpa/src/communication/gossip.rs [hunk 3]
```diff
@@ -1794,4 +1799,55 @@ mod tests {
...
+			_ => panic!("expected no catch up message"),
```
Comment (bkchr) [inline core/finality-grandpa/src/communication/gossip.rs]:
```suggestion
			panic!("expected no catch up message");
```

**Comment by rphmeier** [comment]
Couldn't we ask from anyone advertising that they are a full GRANDPA node watching for votes? I suppose that the main barrier is that we don't have a good way to see that right now.

**Comment by andresilva** [comment]
Yeah, right now we just ask passively when we see a neighbor packet. But in practice we should see a neighbor packet from an authority at some point at which point we'll ask for a catch up.

**Pull request merged**